### PR TITLE
Add support for ingester downscaling

### PR DESCRIPTION
**Pull request by jhalterman**
#### What this PR does

This PR adds support for downscaling ingesters via a new prepare-instance-ring-downscale endpoint. This endpoint sets or clears the read-only status of an ingester in the ring, and will be called by the rollout-operator.

#### Which issue(s) this PR fixes or relates to

Part of https://github.com/grafana/mimir-squad/issues/1797

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/distributor/query.go [hunk 1]
```diff
@@ -137,11 +137,10 @@ func (d *Distributor) getIngesterReplicationSetsForQuery(ctx context.Context) ([
 	// Lookup ingesters ring because ingest storage is disabled.
 	shardSize := d.limits.IngestionTenantShardSize(userID)
 	r := d.ingestersRing
-
-	// If tenant uses shuffle sharding, we should only query ingesters which are part of the tenant's subring.
-	if lookbackPeriod := d.cfg.ShuffleShardingLookbackPeriod; shardSize > 0 && lookbackPeriod > 0 {
-		r = r.ShuffleShardWithLookback(userID, shardSize, lookbackPeriod, time.Now())
+	if shardSize == 0 {
```
Comment (pstibrany) [inline pkg/distributor/query.go:140]: This change is not necessary. We could keep querying entire ring, if shuffle-sharding is not enabled -- entire ring contains all read-only and read-write ingesters, which is fine.

Also setting `shardSize` to `r.InstancesCount()` is not necessary -- `ShuffleShardWithLookback` handles `shardSize == 0` already.

File: pkg/ingester/downscale.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"net/http"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+
+	"github.com/grafana/mimir/pkg/util"
+)
+
+// PrepareInstanceRingDownscaleHandler prepares the ingester ring entry for downscaling. It can mark ingester as READONLY or back to ACTIVE.
```
Comment (pstibrany) [inline pkg/ingester/downscale.go:12]: This entire comment is referring to `READONLY` ring state, but we moved away from that design, and now only mark ingester as read-only, without changing its state.

File: pkg/ingester/downscale.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+//
+// Following methods are supported:
+//
+//   - GET
+//     Returns timestamp when instance was switched to READONLY state, or 0, if partition is not in READONLY state.
```
Comment (pstibrany) [inline pkg/ingester/downscale.go:17]: `PrepareInstanceRingDownscaleHandler` doesn't deal with partitions. (See also comment about `DELETE` method below)

File: pkg/ingester/downscale.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+//
+//   - POST
+//     Switches the instance to READONLY state (if not yet), and returns the timestamp when the switch to
+//     READONLY state happened.
+//
+//   - DELETE
+//     Sets partition back from READONLY to ACTIVE state.
+func (i *Ingester) PrepareInstanceRingDownscaleHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (pstibrany) [inline pkg/ingester/downscale.go:25]: Shall we add a unit test for this handler?

File: pkg/util/atomicfs/fsync.go [hunk 5]
```diff
@@ -68,22 +68,7 @@ func (a *File) Close() error {
...
-	// > needed.
-	dir, err := os.Open(filepath.Dir(a.finalPath))
-	if err != nil {
-		return err
-	}
-
-	merr.Add(dir.Sync())
-	merr.Add(dir.Close())
-	return merr.Err()
+	return syncDir(filepath.Dir(a.finalPath))
```
Comment (pstibrany) [inline pkg/util/atomicfs/fsync.go:71]: Changes in `atomicfs` and `shutdownmarker` packages come from previous design, when we planned to store "downscale marker" on the disk. However we have moved on from that design, and now changes in these two packages are unrelated to this PR.

I suggest we remove changes to `atomicfs` and `shutdownmarker` packages from this PR.

Comment (pstibrany) [reply pkg/util/atomicfs/fsync.go]: Done in https://github.com/grafana/mimir/pull/8956/commits/0d64c249f3461cce83c8f3a730fdbdb0abe1ef8e.

File: pkg/ingester/downscale.go (hunk 3 above)
Comment (pstibrany) [reply pkg/ingester/downscale.go]: Fixed in https://github.com/grafana/mimir/pull/8956/commits/96afdc1c3d2f62572d1e7f296b0f96552980a9df.

File: pkg/ingester/downscale.go (hunk 2 above)
Comment (pstibrany) [reply pkg/ingester/downscale.go]: Updated in https://github.com/grafana/mimir/pull/8956/commits/96afdc1c3d2f62572d1e7f296b0f96552980a9df.

File: pkg/ingester/downscale.go (hunk 4 above)
Comment (pstibrany) [reply pkg/ingester/downscale.go:28]: I started in https://github.com/grafana/mimir/pull/8956/commits/9bb33466be1ee28abebd65fcd91f4c3cfb1fd328, but it needs more work.

Comment (jhalterman) [reply pkg/ingester/downscale.go:28]: Added some more test cases

**Review by pstibrany — APPROVED** [review]
Thank you! I've pushed some comments and test changes, but nothing substantial, so I'm approving this.
